google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelector.java

```diff
@@ -0,0 +1,220 @@
+  }
+
+  /**
+   * Adds a {@link ProjectSelectionListener} to this class's internal list of listeners. All
+   * ProjectSelectionListeners are notified when the selection is changed to a valid project.
+   *
+   * @param projectSelectionListener the listener to add
+   */
+  public void addProjectSelectionListener(ProjectSelectionListener projectSelectionListener) {
+    projectSelectionListeners.add(projectSelectionListener);
+  }
+
+  /**
+   * Removes a {@link ProjectSelectionListener} from this class's internal list of listeners.
+   *
+   * @param projectSelectionListener the listener to remove
+   */
+  public void removeProjectSelectionListener(ProjectSelectionListener projectSelectionListener) {
+    projectSelectionListeners.remove(projectSelectionListener);
+  }
+
+  private void initUi() {
+    // layout - in the center of panel, horizontal line of project/account labels in a panel
+    // ends with a fixed size browse button.
+    setLayout(new BorderLayout());
+
+    JPanel staticInfoPanel = new JPanel();
+    staticInfoPanel.setLayout(new FlowLayout(FlowLayout.LEFT, 5, 2));
+    staticInfoPanel.setBorder(UIManager.getBorder("TextField.border"));
+
+    projectNameLabel = new HyperlinkLabelWithStateAccess();
+    projectNameLabel.setHyperlinkText(GctBundle.getString("project.selector.no.selected.project"));
+    projectNameLabel.addHyperlinkListener(
+        (event) -> {
+          if (event.getEventType() == EventType.ACTIVATED) {
+            handleOpenProjectSelectionDialog();
+          }
+        });
+    staticInfoPanel.add(projectNameLabel);
+
+    projectAccountSeparatorLabel = new JBLabel("/");
+    projectAccountSeparatorLabel.setVisible(false /* only visible when project is selected. */);
+    staticInfoPanel.add(projectAccountSeparatorLabel);
+
+    accountInfoLabel = new HyperlinkLabelWithStateAccess();
+    accountInfoLabel.addHyperlinkListener(
+        (event) -> {
+          if (event.getEventType() == EventType.ACTIVATED) {
+            handleOpenProjectSelectionDialog();
+          }
+        });
+    staticInfoPanel.add(accountInfoLabel);
+
+    ComponentWithBrowseButton<JPanel> componentWithBrowseButton =
+        new ComponentWithBrowseButton<>(
+            staticInfoPanel, (actionEvent) -> handleOpenProjectSelectionDialog());
+
+    add(componentWithBrowseButton);
+  }
+
+  @VisibleForTesting
+  void handleOpenProjectSelectionDialog() {
+    ProjectSelectionDialog projectSelectionDialog = createProjectSelectionDialog(this);
+    CloudProject newSelection = projectSelectionDialog.showDialog(cloudProject);
+
+    // if null, it means no change or user cancelled selection dialog - no update required.
+    if (newSelection != null) {
+      setSelectedProject(newSelection);
+      notifyProjectSelectionListeners();
+    }
+  }
+
+  private void updateProjectAndUserInformation(CloudProject selection) {
+    if (selection == null || Strings.isNullOrEmpty(selection.googleUsername())) {
```

Comment:
maybe consider splitting up this method for readability? I.e. so that it internally calls something like `updateEmptySelection`, and `updateCloudProjectSelection` (in fact you could probably move this out to the caller of this method if you decide to take this suggestion).

Reply:
Probably better to split, since now it technically does two different things instead of clear one.